### PR TITLE
feat(mcp): honor resources/subscribe by polling GitLab (ADR-0015)

### PR DESCRIPTION
**Pull request by jmrplens**
## Description

Honors `resources/subscribe` by polling GitLab, under bounds an operator can predict. This reverses [ADR-0010](docs/development/adr/adr-0010-no-resource-subscribe.md), which declined the capability partly because no client had asked for it.

That premise turned out to be wrong, and the way it was wrong is the point: **VS Code subscribes to every resource it reads**, and **Cursor sends `resources/subscribe` even to servers advertising `subscribe: false`**. Declining never stopped those subscriptions from being made — it only guaranteed they failed silently. The remaining objection in ADR-0010 was API budget, which is a number, and a number can be bounded.

[ADR-0015](docs/development/adr/adr-0015-polled-resource-subscriptions.md) records the decision and its bounds; ADR-0010 keeps its reasoning with a note pointing at what changed.

## Related Issue

N/A

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Enhancement (improvement to existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Refactoring (no functional change)
- [x] Documentation update
- [ ] CI / build / tooling

## Changes Made

- **`internal/subscriptions/`** (new) — a whitelist of 26 single-object resource kinds, polled watchers, cadence, lease and rate-limit back-off. Knows nothing about MCP.
- **`cmd/server/subscriptions.go`** (new) — the bridge: reader, notifier, session bridge, renewal middleware, graceful stream close.
- **`internal/resources/registrar.go`** (new) — a seam so the server can read its own resources through the very handlers `resources/read` dispatches to, rather than a second read path that would drift from them.
- **`internal/config`** — `Stateless` reaches `ServerConfig`, because the transport decides whether the legacy subscribe path can be honored at all.

### Three decisions worth reviewing, because the obvious alternative is wrong in each

1. **The lease demotes a watch to a slow poll; it never stops one.** MCP defines no lease, no TTL and no expiry message, and its one notification means "this changed, read it again". A watch that retired at its deadline would go silent in a way no client could tell apart from "nothing has happened yet". Demotion is the only claim the protocol lets a server make honestly.
2. **Nothing retires itself for reaching a terminal state, because GitLab has none.** Retrying a pipeline reuses its ID and starts it running again; a closed issue reopens; a merged MR can still be edited. A watcher that stopped at `success` would go blind to a real change its subscriber asked about.
3. **A subscriber is an identity, not a count.** With a refcount, one session subscribing twice leaves a hold nothing can release, and an unsubscribe from a session that never subscribed stops somebody else's watch.

### Bounds

| | |
| --- | --- |
| Subscribable | 26 single-object kinds; collections deliberately excluded |
| Cadence | 5s floor while work is in flight, 15s default, 60s once settled |
| Watchers | 10 per server, and a server is one token |
| Lease | 30 min → 10-minute poll; renewed by any request that reaches the server |
| Hard stop | 24h absolute cap, a 401/403/404, eviction at the cap, unsubscribe, or session end |
| Rate limit | one 429 pauses every watcher on the manager, 30s doubling to 5 min |

Ten watchers at the 15s base cost 40 req/min; the same ten demoted cost one. A self-managed instance's default is 120 req/min, and that limit is off unless an administrator enables it.

## How to Test

1. `go test ./internal/subscriptions/ ./internal/resources/ ./cmd/server/ -count=1`
2. `make analyze` — all six gates.
3. End to end, the claim the feature rests on: `go test ./cmd/server/ -run TestSubscribe_ResourceChanges_ClientIsNotified -count=1 -v` — a real client subscribes, the mock GitLab's pipeline status changes, and the notification arrives over the wire.
4. The graceful close, which is the part with no precedent in the Go SDK: `go test ./cmd/server/ -run TestSubscribe_WatchStops_StreamEndsWithAResult -count=1 -v`.

## Breaking Changes / Migration Notes

N/A for existing clients — nothing changes for one that does not subscribe. The capability is advertised only on `CAPABILITY_SURFACE=full` (the default), where the resources it applies to are registered.

One behavior worth knowing: **in stateless HTTP mode (the default), the legacy `resources/subscribe` is refused with an explanatory error.** The SDK gives each stateless POST its own session and closes it when the POST returns, so such a subscription would be cancelled moments after being acknowledged, having already spent a GitLab round-trip on the authorization read. `subscriptions/listen` (protocol 2026-07-28) is unaffected, because there the subscription *is* an open request. Run with `--stateless=false` to serve legacy subscribers over HTTP.

### Code Quality

- [x] Code compiles: `go build ./...`
- [x] Consolidated Go analysis passes on changed packages (or `make golangci-lint` / `make analyze`)
- [x] Code is formatted: `make analyze-fix`
- [x] Follows idiomatic Go patterns and the conventions documented in `CLAUDE.md` / `.github/instructions/`

### Testing

- [x] All existing tests pass: `go test ./... -count=1`
- [x] New tests added for new functionality (table-driven, with `httptest` mocks)
- [x] Coverage on modified packages is ≥ 90%: `subscriptions` 98.1%, `resources` 98.2%, `config` 99.5%
- [x] Edge cases and error scenarios covered
- [ ] If applicable, E2E tests updated/added under `test/e2e/suite/` — not applicable: the e2e suite drives tools, and a subscription needs a client that opens one

### Documentation

- [x] Doc comments added for exported types/functions (godoc)
- [x] `docs/` updated: ADR-0015, ADR-0010 superseded, capability reference, capabilities index, resources reference
- [x] `README.md` regenerated (stats)
- [ ] If introducing/removing a tool: not applicable — this adds a capability, not a tool
- [x] Commit messages follow Conventional Commits

### Security

- [x] No secrets, tokens, or credentials in code, tests, fixtures, or logs
- [x] Input validation: the URI whitelist parses segment by segment and refuses anything the resource router could not resolve
- [x] Error messages do not leak sensitive information — 401, 403 and 404 collapse to one sentinel, matching GitLab's own refusal to distinguish "absent" from "not yours"
- [x] No new dependencies

## Review notes

A 38-agent adversarial review ran against this branch before it was opened. It raised 33 findings; 20 survived independent verification and all are fixed here. The ones that changed the design rather than a line are the three decisions listed above, plus:

- **Eviction at the watcher cap could never fire.** The renewal middleware ran *before* the handler, so a subscribe request un-demoted every watcher a fraction of a second before that same request looked for a demoted one to evict. Renewal now runs after the handler.
- **`mcp.ResourceNotFoundError` was not recognised as inaccessible**, so a deleted repository file kept being polled for a full day instead of stopping.
- **Labels addressed by name** were readable but rejected as not subscribable, breaking the whitelist's own invariant that it mirrors what the resource router can resolve.

One correction to note honestly: an earlier revision of this branch silently deleted the pre-existing `KeepAlive: 30 * time.Second` from `mcp.ServerOptions` to make a flaky HTTP test pass. That was wrong, it is restored, and the flakiness was then measured against a clean worktree at `main` — where those same HTTP tests fail identically under `-race`. It is pre-existing and queued as its own change.

<a href="https://cubic.dev/pr/jmrplens/gitlab-mcp-server/pull/296?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Added polling-based subscriptions for 26 supported single-object resource types.
  * Sends notifications when subscribed resource content changes.
  * Supports adaptive polling, subscription streams, lease renewal, and automatic cleanup.
  * Advertises subscriptions only on the full capability surface.
* **Bug Fixes**
  * Improved handling of unsupported, inaccessible, rate-limited, expired, and closed subscriptions.
  * Stateless HTTP now directs supported clients to `subscriptions/listen`.
  * Added guidance to watch in-progress pipeline resources instead of repeatedly polling.
* **Documentation**
  * Added subscription reference, compatibility, configuration, and troubleshooting guidance.

### CODE REVIEW

File: CLAUDE.md [hunk 1]
```diff
@@ -566,6 +568,21 @@ Plus enterprise-only routes injected into 3 base meta-tools:
...
+- **A joiner waits for the first read** — the watcher's `ready` channel carries the outcome of the initial read (which is the authorization check), so a second subscriber is never told "subscribed" while the only read anyone attempted is still in flight or has already failed.
+- **Renewal happens after the handler, not before it** — a subscribe request is itself activity, so renewing first would un-demote every watcher a moment before that same request looked for a demoted one to evict, and eviction at the cap could never fire.
+- **The lease demotes, it does not stop** — 30 minutes without traffic drops a watch to a 10-minute poll; any request that reaches the server restores it (`renewOnActivity` middleware). From protocol 2026-07-28 a client may answer `tools/list` or a repeated `resources/read` from its own cache, so those never renew. Only `MaxLifetime` (24h), a 401/403/404, or eviction at the cap actually stop one.
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply CLAUDE.md]: Applied in c1a9373f. The sentence now separates the two directions: `MaxLifetime`, a 401/403/404, and cap eviction stop a watch without the client acting; `resources/unsubscribe` and session disconnect stop one from the client side.

File: docs/reference/capabilities/subscriptions.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+on a project, and a single job; a merge request with its discussions and
+notes; an issue; a deployment, an environment, a feature flag; a release, a
+tag, a branch; a milestone (project or group), a label (project or group), a
+board; a deploy key; a snippet (project or personal); a wiki page; and a
+repository file.
+
+**Collections are deliberately excluded.** Subscribing to
+`gitlab://project/42/issues` would notify on every change to any issue in
+the project — most of them nothing the subscriber asked about — and cost a
+full page read on every poll. A subscription to a collection is refused.
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply docs/reference/capabilities/subscriptions.md]: Partly disagree on the facts, but the wording deserved fixing — done in c1a9373f.

The three list-shaped entries are in the whitelist deliberately: `internal/subscriptions/kind.go` defines `KindPipelineJobs`, `KindMergeRequestDiscussions` and `KindMergeRequestNotes`, and `Manager.Subscribe` accepts them. They are bounded, single-parent lists — a pipeline's jobs, one MR's discussion set — whose parent's lifecycle is exactly what the subscriber is following. What the classifier refuses is the open-ended, top-level collection (`gitlab://project/42/issues`), where changes arrive for reasons the subscriber never asked about.

What was genuinely wrong was this doc calling all 26 kinds "single objects". The paragraph now names the three attached lists explicitly and scopes the refusal to top-level collections, so the doc and the classifier agree.

File: internal/subscriptions/kind.go [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+	// Snippets are addressed by a numeric ID; projects and groups accept
+	// either a numeric ID or an encoded path.
+	if prefix == snippetPrefix && !isPositiveInt(ref) {
+		return KindUnknown, false
+	}
+
+	if tailPath == "" {
+		kind := bareKinds[prefix]
+		return kind, kind != KindUnknown
+	}
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply internal/subscriptions/kind.go]: Confirmed and fixed in c1a9373f: `Classify` now rejects a separator with an empty tail — `gitlab://project/42/`, `gitlab://group/9/` and `gitlab://snippet/5/` all return `KindUnknown, false` — with those exact cases added to `TestClassify_MalformedURI_IsRejected`.

One correction to the failure scenario: the outcome would not have been "acknowledged then dropped". `Manager.Subscribe` performs the first read synchronously on the caller's context before accepting the subscription — that read is the acceptance gate — so an unresolvable URI is refused up front with an error. Still worth fixing here, because the whitelist's stated invariant is that it mirrors what the router can resolve, and the trailing-slash shape broke that invariant.

File: internal/subscriptions/readerror.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+		return ErrRateLimited
+	case toolutil.IsHTTPStatus(err, http.StatusUnauthorized),
+		toolutil.IsHTTPStatus(err, http.StatusForbidden),
+		toolutil.IsNotFound(err),
+		isResourceNotFound(err):
+		return ErrInaccessible
+	default:
+		return err
+	}
+}
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply internal/subscriptions/readerror.go:48]: Applied in c1a9373f. `TranslateReadError` now wraps both sentinels around the original error (`%w: %w`), the inaccessible-stop log line and the rate-limit warning carry the underlying error, and `retire` receives the wrapped error — `errors.Is` matching is unchanged throughout.

File: .plugin/plugin.json [hunk 5]
```diff
@@ -2,7 +2,7 @@
   "$schema": "./plugin.schema.json",
   "name": "gitlab-mcp-server",
   "version": "2.6.6",
-  "description": "GitLab MCP server exposing 1027 self-managed GitLab REST/GraphQL MCP tools, or 1033 on GitLab.com Enterprise/Premium with Orbit. Includes a default 2-tool dynamic surface, 33 base / 49 self-managed enterprise / 50 GitLab.com Enterprise meta-tools, 46 resources, 37 prompts, 17 completion types, and 6 MCP capabilities (logging, progress, roots, sampling, elicitation, completions).",
+  "description": "GitLab MCP server exposing 1027 self-managed GitLab REST/GraphQL MCP tools, or 1033 on GitLab.com Enterprise/Premium with Orbit. Includes a default 2-tool dynamic surface, 33 base / 49 self-managed enterprise / 50 GitLab.com Enterprise meta-tools, 46 resources, 37 prompts, 17 completion types, and 4 MCP capabilities (completions, progress, elicitation, resource subscriptions).",
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply .plugin/plugin.json]: Applied in c1a9373f — `45 resources`, matching the generated artifacts (`site/src/data/stats.json` and the `llms*.txt` files, both of which derive the count from the live registration).

File: cmd/server/instructions.go [hunk 6]
```diff
@@ -160,5 +163,14 @@ func buildInstructions(toolSurface string) string {
 		"2. ID is the global numeric identifier. Only use %s when you have a global ID from another API response.",
 		refIssueGetByID.render(toolSurface))
 
+	if capabilitySurface == config.CapabilitySurfaceFull {
+		b.WriteString("\n\nWATCHING RESOURCES — Instead of re-reading a resource in a loop to detect change:\n" +
+			"1. Single-object resources (a pipeline, an issue, a merge request, a file, a wiki page, ...) can be " +
+			"watched for change notifications via MCP resources/subscribe; collections (issue lists, branch lists) cannot.\n" +
+			"2. Example: subscribe to gitlab://project/{project_id}/pipelines/latest to be notified when a pipeline's " +
+			"state changes, instead of polling it yourself — the server watches GitLab and sends " +
+			"notifications/resources/updated only when the content actually changed.")
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply cmd/server/instructions.go]: Applied in c1a9373f. `buildInstructions` now takes the stateless flag from `ServerConfig`: on stateless HTTP the WATCHING RESOURCES section names `subscriptions/listen` and states that the legacy `resources/subscribe` is refused on that transport; elsewhere it keeps naming `resources/subscribe`. Added `TestBuildInstructions_StatelessHTTP_NamesTheWorkingMethod` covering the full-capability × stateless combination in both directions (names the working method, never teaches the refused one).

File: docs/concepts/resource-consumption.md [hunk 7]
```diff
@@ -78,19 +79,40 @@ CPU usage depends on request throughput, not pool size:
 | Per active MCP session           | ~2         |
 | Per concurrent tool call         | 1          |
 
+| Per active resource subscription | 1         |
+
 A server with 100 active sessions and 10 concurrent tool calls: ~220 goroutines total.
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply docs/concepts/resource-consumption.md]: Applied in c1a9373f. The goroutine table now counts per distinct watched URI (max 10 per server) with a note that subscriptions to the same URI join one watcher, the prose says "starts — or joins — a watcher", and the example total includes the ten watchers (~230).

File: docs/guides/client-compatibility.md [hunk 8]
```diff
@@ -40,6 +40,17 @@ These are client-side constraints, not server behavior. The default `dynamic` su
 | Gemini CLI                                                                  | Tool names over 63 characters truncated; Gemini API rejects `$ref`/`$defs` in input schemas                                                                                                                                                                                                        |
 | JetBrains AI Assistant                                                      | Rejects the whole `tools/list` if any tool's `outputSchema` root type is not `object` ([LLM-30555](https://youtrack.jetbrains.com/issue/LLM-30555))                                                                                                                                                |
 
+## Subscription behavior per client
+
+Client handling of `resources/subscribe` varies more than any other
+capability: VS Code subscribes to every resource it reads and routes
+`resources/updated` into its file-change pipeline (not into chat); Cursor
+sends `resources/subscribe` even to servers advertising `subscribe: false`;
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply docs/guides/client-compatibility.md:48]: Applied in c1a9373f: "VS Code attempts to subscribe to every resource it reads (this server accepts only the 26 subscribable kinds and refuses the rest)".

File: docs/guides/troubleshooting.md [hunk 9]
```diff
@@ -121,6 +121,31 @@ See [HTTP Server Mode — OAuth Mode](http-server-mode.md#oauth-mode) for the fu
...
+(protocol 2026-07-28) works there. On protocol 2026-07-28 a refusal may
+never reach the client (the Go SDK fires the listen request without awaiting
+its response), so a subscription that "succeeded" but never fires is often
+one that was refused.
+
+**Notifications arrive slowly.** The watch may have lease-demoted: 30
+minutes without any request on the session drops it to a 10-minute poll.
+Any tool call or resource read on that session restores full speed. The
+notification's `_meta` (`io.github.jmrplens/watch`) reports the current
+state and cadence.
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply docs/guides/troubleshooting.md]: Applied in c1a9373f. The troubleshooting entry now distinguishes the two slowdowns — settled resources poll at 60s by design (BaseInterval × settledFactor), lease demotion drops to the 10-minute SlowInterval — and `TestWatchMeta_DescribesTheWatch` gained a settled-cadence case asserting `state: "active"` with `pollIntervalMs: 60000`, so the `_meta` lets a client tell the two apart.

File: internal/prompts/prompts.go [hunk 10]
```diff
@@ -330,6 +330,17 @@ func handleSummarizePipelineStatus(ctx context.Context, client *gitlabclient.Cli
...
+	case "success", "failed", "canceled", "skipped", "manual":
+		// Settled: nothing to wait for.
+	default:
+		// Still moving. Pointing at the subscription beats teaching the
+		// model to re-invoke this prompt on a timer, which is the pattern
+		// resource subscriptions exist to replace.
+		fmt.Fprintf(&b, "This pipeline is still in progress — if your client supports MCP resource "+
+			"subscriptions, subscribe to gitlab://project/%s/pipelines/latest to be notified when it "+
+			"changes instead of re-invoking this prompt.\n", projectID)
+	}
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply internal/prompts/prompts.go:345]: Applied in c1a9373f. `scheduled` joined the settled cases — same reasoning the polling cadence already encodes: both `manual` and `scheduled` wait on something outside CI. Added table-driven `TestSummarizePipelineStatusPrompt_SubscriptionHint` covering eight statuses in both directions (hint present for running/pending/created/waiting_for_resource, absent for success/failed/manual/scheduled).

File: internal/resources/workflow_guides.go [hunk 11]
```diff
@@ -237,6 +237,13 @@ BREAKING CHANGE: /auth/login is now /api/v2/auth/login
 4. Check variables: ` + "`gitlab_ci_variable_list`" + ` for missing configuration.
 5. Retry the job: ` + "`gitlab_job_retry`" + ` for transient failures.
 
+## Watching Instead of Polling
+If your client supports MCP resource subscriptions (resources.subscribe),
+subscribe to ` + "`gitlab://project/{project_id}/pipelines/latest`" + ` (or
+` + "`gitlab://project/{project_id}/pipeline/{pipeline_id}`" + ` for a specific
+pipeline) instead of re-checking status in a loop — the server polls GitLab
+for you and sends a resources/updated notification when the pipeline changes.
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply internal/resources/workflow_guides.go:246]: Applied in c1a9373f. The guide now reads "resources.subscribe — or subscriptions/listen from protocol 2026-07-28, the only form stateless HTTP honors".

File: internal/subscriptions/kind_test.go [hunk 12]
```diff
@@ -0,0 +1,743 @@
...
+	want := make(map[string]bool, len(kindTemplates))
+	for _, template := range kindTemplates {
+		want[template] = true
+	}
+	for _, template := range got {
+		if !want[template] {
+			t.Errorf("Templates() advertises %q, which the whitelist does not contain", template)
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply internal/subscriptions/kind_test.go:747]: Declining this one. The flagged tests are single-scenario assertions — `TestTemplates_MatchesWhitelistExactly` checks one property of one function result, and a one-row table would add indirection without adding cases. The multi-case tests beside them (`TestKindCount_MatchesEveryDocumentThatCitesIt`, `TestClassify_*`) already use `t.Run` tables per the repository convention. Happy to convert if these ever grow a second scenario.

File: server.json [hunk 13]
```diff
@@ -72,7 +72,7 @@
         },
         {
           "name": "CAPABILITY_SURFACE",
-          "description": "Resource and prompt catalog selector: full or minimal (default: full)",
+          "description": "Resource and prompt catalog selector: full or minimal (default: full); resource subscriptions require full",
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply server.json:75]: This is working as designed, so no change here — the in-repo `server.json` is not what clients consume. The MCP Registry only receives it at release time, when the tag-triggered workflow runs `scripts/update-server-json-sha.sh` to stamp version, URLs and `fileSha256` for the release being published (see the release process in CLAUDE.md). The env-var description edited in this PR ships to the registry in the same publish that carries the subscriptions feature, so there is no window in which the registry advertises subscriptions against v2.6.6 artifacts. The MCPB link follows the same tag-time stamping.

File: site/src/content/docs/capabilities/overview.mdx [hunk 14]
```diff
@@ -1,34 +1,36 @@
...
+| [Progress](/gitlab-mcp-server/capabilities/progress/)           | Server → Client                           | Real-time progress updates for long-running operations                                                      |
+| [Elicitation](/gitlab-mcp-server/capabilities/elicitation/)     | Server → Client                           | Interactive wizards — step-by-step forms for creating complex resources                                     |
+| [Completions](/gitlab-mcp-server/capabilities/completions/)     | Client → Server                           | Argument autocompletion for project names, branches, users, and more                                        |
+| [Subscriptions](/gitlab-mcp-server/capabilities/subscriptions/) | Client → Server (updates Server → Client) | Live change notifications for 26 GitLab resource kinds, honored by polling (`CAPABILITY_SURFACE=full` only) |
+| [Icons](/gitlab-mcp-server/capabilities/icons/) _(metadata)_    | Server → Client                           | Icons for every tool, resource, and prompt — SVG plus light/dark WebP fallbacks                             |
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply site/src/content/docs/capabilities/overview.mdx:33]: Applied in c1a9373f. The matrix gained a `Subscriptions` row in both languages, with a footnote stating exactly what is verified: VS Code subscribes automatically and routes `resources/updated` into its file-change pipeline; Cursor sends `resources/subscribe` requests but whether it surfaces notifications is unverified, as is handling in Claude Desktop and Claude Code; clients that never subscribe lose nothing.

File: site/src/content/docs/es/operations/http-server.mdx [hunk 15]
```diff
@@ -219,6 +219,8 @@ gitlab-mcp-server --http --gitlab-url=https://gitlab.example.com \
 
 ### Modo con estado heredado
 
+En modo sin estado la petición legada `resources/subscribe` se rechaza con un error explicativo — cada POST recibe su propia sesión que se cierra con la respuesta, así que una suscripción aceptada nunca podría notificarse; los clientes con protocolo 2026-07-28 conservan las [suscripciones a recursos](/gitlab-mcp-server/es/capabilities/subscriptions/) mediante `subscriptions/listen`, que mantiene la petición abierta.
+
 `--stateless=false` restaura el transporte basado en sesiones: se emite `Mcp-Session-Id` en `initialize`, GET abre el flujo SSE independiente, DELETE termina la sesión y `--session-timeout` gobierna el tiempo de vida en reposo. Es un modo de compatibilidad para clientes que aún no negocian el protocolo `2026-07-28` (negocian `2025-11-25` o anterior y usan elicitación síncrona), el servidor registra un aviso al arrancar cuando está activo, y la intención es retirarlo cuando los ecosistemas de clientes hayan migrado.
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply site/src/content/docs/es/operations/http-server.mdx]: Applied in c1a9373f — the paragraph now sits above the heading, in its stateless context. The EN twin had the identical placement defect and was fixed in the same commit.

File: internal/prompts/prompts_test.go [hunk 16]
```diff
@@ -202,6 +202,57 @@ func TestSummarizePipelineStatusPrompt_Success(t *testing.T) {
...
+		wantHint bool
+	}{
+		{"running", true},
+		{"pending", true},
+		{"created", true},
+		{"waiting_for_resource", true},
+		{"success", false},
+		{"failed", false},
+		{"manual", false},
+		{"scheduled", false},
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply internal/prompts/prompts_test.go:226]: Applied in f1c5ef1d — `canceled` and `skipped` added with `wantHint: false`, so every status the production switch names as settled has its own regression case.

File: internal/serverpool/pool_test.go [hunk 17]
```diff
@@ -993,7 +1016,7 @@ func TestGetOrCreate_EmptyGitLabURL(t *testing.T) {
 // TestGetOrCreate_DifferentURLsSameToken verifies that the same token
 // against different GitLab instances gets separate pool entries.
 func TestGetOrCreate_DifferentURLsSameToken(t *testing.T) {
-	cfg := testConfig("http://localhost")
+	cfg := testConfig(stubGitLabBase)
 	pool := New(cfg, testFactory())
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply internal/serverpool/pool_test.go:1020]: Applied in f1c5ef1d. The test now stands up two local `httptest` stubs and passes their URLs, with a comment recording why: `GetOrCreate` probes the URL it is given, so the example.com hosts meant real DNS lookups from a unit test that passed only because the probe fails open.

File: internal/subscriptions/manager.go [hunk 18]
```diff
@@ -0,0 +1,897 @@
...
+
+// Manager owns the watchers for one MCP server, which in HTTP mode means
+// one GitLab token: the server pool keys an *mcp.Server by token and URL,
+// so a manager never spans two identities and a rate-limit pause or an
+// access revocation applies cleanly to everything it owns.
+type Manager[S comparable] struct {
+	reader   Reader
+	notifier Notifier
+	opts     Options
+	wg       sync.WaitGroup
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply internal/subscriptions/manager.go:217]: Applied in f1c5ef1d: `m.wg.Add(1)` moved inside the critical section that proves the manager open, with a comment stating the invariant.

One scoping note on the scenario: the escaped goroutine's context was already cancelled — `Close` collects `w.cancel` under the same lock that made the watcher visible — so it exited on its first select without ever polling. What the window actually broke was `Close`'s "waits for them to finish" contract, which held only by that accident. For the barrier test: the interleaving cannot be forced deterministically without injection hooks, so `TestClose_RacesSubscribe_NeverOrphansAWatcher` hammers it instead — 100 iterations of a concurrent Subscribe racing Close at a 1ms cadence under `-race`, asserting that once Close returns, nothing reads again.

File: internal/subscriptions/manager_test.go [hunk 19]
```diff
@@ -0,0 +1,1467 @@
...
+		wg.Wait()
+
+		// Close has returned: however the race fell, nothing may read
+		// again. Two poll periods of quiet is proof at this cadence.
+		settled := r.readCount(testURI)
+		time.Sleep(5 * time.Millisecond)
+		if got := r.readCount(testURI); got != settled {
+			t.Fatalf("reads went %d -> %d after Close returned; a watcher outlived shutdown", settled, got)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply internal/subscriptions/manager_test.go:822]: Partially applied in 7d3a05ee, with the conversion itself declined — reasoning below, since the distinction is worth recording.

These 100 passes are not a case table: every iteration runs the identical scenario, and the number exists only to give the race detector enough interleavings to explore. The table-driven convention this repository holds (and the sibling tests in this file follow) is about enumerating *distinct* cases with meaningful names; a table of 100 identical entries named 0..99 would misstate what the test is. The Go standard library uses plain repetition loops for exactly this kind of interleaving hunt.

What the comment is right about is diagnosability on failure, so that part is taken: the failure message now carries the iteration index that tripped.

File: internal/resources/registrar_test.go [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+	}
+
+	blob, err := index.Read(context.Background(), "tmpl/blob", "gitlab://x/1")
+	if err != nil {
+		t.Fatalf("Read(blob): %v", err)
+	}
+	if len(blob) != 2 || blob[0] != 0x1f {
+		t.Errorf("Read(blob) = %v, want the handler's blob bytes", blob)
+	}
+}
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply internal/resources/registrar_test.go:139]: Applied in a965e0a6 — the two content cases are now named subtests over a table, so a text failure cannot mask the blob case or vice versa.

File: internal/subscriptions/kind_test.go [hunk 21]
```diff
@@ -0,0 +1,747 @@
...
+
+// TestKindString_EveryKind_HasADistinctName verifies each kind renders a
+// distinct, non-empty name, since these strings appear in logs and
+// rejection messages where "unknown" for a real kind would mislead.
+func TestKindString_EveryKind_HasADistinctName(t *testing.T) {
+	seen := make(map[string]Kind, len(kindMeta))
+	for k, meta := range kindMeta {
+		name := meta.name
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply internal/subscriptions/kind_test.go]: Applied in a965e0a6, and thanks for this one — it was a real regression in the test's meaning. The kindMeta refactor had quietly turned the assertion into checking the backing table against itself; a String() that collapsed every mapped kind to one constant would have sailed through. The loop calls `k.String()` again.

File: internal/subscriptions/kind.go [hunk 22]
```diff
@@ -0,0 +1,376 @@
...
+	templates := make([]string, 0, len(kindMeta))
+	for _, meta := range kindMeta {
+		templates = append(templates, meta.template)
+	}
+	slices.Sort(templates)
+	return templates
+}
+
+func (k Kind) Template() string {
+	return kindMeta[k].template
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply internal/subscriptions/kind.go:145]: Applied in a965e0a6 — the kindMeta merge had swallowed `Template`'s doc comment when `Templates` moved in above it; restored.
